core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -526,42 +697,92 @@ private void abortBlockInternal(long sessionId, long blockId) throws BlockDoesNo
+      tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+      srcPath = tempBlockMeta.getPath();
+      dstPath = tempBlockMeta.getCommitPath();
+      loc = tempBlockMeta.getBlockLocation();
+    }
+
+    // Heavy IO is guarded by block lock but not metadata lock. This may throw IOException.
+    FileUtils.move(srcPath, dstPath);
+
+    try (LockResource r = new LockResource(mMetadataWriteLock)) {
+      mMetaManager.commitTempBlockMeta(tempBlockMeta);
+    } catch (BlockAlreadyExistsException | BlockDoesNotExistException
+        | WorkerOutOfSpaceException e) {
+      throw Throwables.propagate(e); // we shall never reach here
+    }
+
+    // Check if block is pinned on commit
+    if (pinOnCreate) {
+      addToPinnedInodes(BlockId.getFileId(blockId));
+    }
+
+    return loc;
+  }
+
+  private StorageDirView allocateSpace(long sessionId, AllocateOptions options) {
+    StorageDirView dirView = null;
+    BlockMetadataView allocatorView = new BlockMetadataAllocatorView(mMetaManager);
     try {
-      // When committing TempBlockMeta, the final BlockMeta calculates the block size according to
-      // the actual file size of this TempBlockMeta. Therefore, commitTempBlockMeta must happen
-      // after moving actual block file to its committed path.
-      BlockStoreLocation loc;
-      String srcPath;
-      String dstPath;
-      TempBlockMeta tempBlockMeta;
-      try (LockResource r = new LockResource(mMetadataReadLock)) {
-        checkTempBlockOwnedBySession(sessionId, blockId);
-        tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
-        srcPath = tempBlockMeta.getPath();
-        dstPath = tempBlockMeta.getCommitPath();
-        loc = tempBlockMeta.getBlockLocation();
+      // Allocate from given location.
+      dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+          options.getLocation(), allocatorView);
+
+      if (dirView != null) {
+        return dirView;
       }
 
-      // Heavy IO is guarded by block lock but not metadata lock. This may throw IOException.
-      FileUtils.move(srcPath, dstPath);
+      if (options.isForceLocation()) {
+        if (options.isEvictionAllowed()) {
+          freeSpace(sessionId, options.getSize(), options.getSize(), options.getLocation());
+          dirView = mAllocator.allocateBlockWithView(sessionId, options.getSize(),
+              options.getLocation(), allocatorView.refreshView());
 
-      try (LockResource r = new LockResource(mMetadataWriteLock)) {
-        mMetaManager.commitTempBlockMeta(tempBlockMeta);
-      } catch (BlockAlreadyExistsException | BlockDoesNotExistException
-          | WorkerOutOfSpaceException e) {
-        throw Throwables.propagate(e); // we shall never reach here
-      }
+          if (dirView == null) {
+            LOG.error("Target tier: {} has no evictable space to store {} bytes for session: {}",
+                options.getLocation(), options.getSize(), sessionId);
```

Comment:
Should we log this, I don't think its unexpected? Maybe we can increment some metric to keep track of how many failed allocations we had.